Amend RFC 92 With New Options (#99)
getsentry/rfcs: text/0092-replay-issue-creation.md

**Pros:**

- Significantly larger customer reach (i.e. all javascript SDK customers).
- Opportunity to upsell customers on Session Replay from a "slow click" event.
- No code changes required on the Session Replay back end.
- We can sample slow click issues on the backend.
  - This would require buy-in and coordination from the Issues team.
  - There is nothing preventing us from looking at an HTTP request and making a go/no-go decision.

**Cons:**

- Requires SDK changes.
- Requies code changes by the Issues team to create a generic interface for creating issues.
  - This would likely disrupt our June 18 deadline.
  - Unless the Issues team has excess capacity and a willingness to work on it immediately.

### Option 3: Replay SDK Pushes Issues to Session Replay Backend Which Raises an Issue

When the SDK encounters a "slow click" it will append the slow click to the recording events payload. The back-end will search for these events and then publish them to the Issues platform's kafka consumer.

**Pros:**

- We can sample slow click events on the back-end without worrying about input from other teams.

> - No code changes required by the Issues team.

Reviewer: ```suggestion
- No code changes required by the Issues team.
- May be able to incentivize folks to install Replay in order to see dead click issues
```

**Cons:**

- Requires the Session Replay SDK.
  - Significantly smaller pool of customers who will see "slow click" issues.
- Requires code changes by the Session Replay back-end team.
  - Requires addition of event sampling, issue platform integration, and merging of the replay-event and recording-event payloads.
  - Merging the replay-event and recording-event payloads together is not a trivial change and requires careful deployment.

> # Unresolved questions

Reviewer 2: We should put both of these bullets as Cons inside (especially) option 1

# Options Considered

> ### Option 1: SDK Creates Issues Through the Error Interface

Reviewer 3: How about just:
```suggestion
### Option 1: SDK Creates Issues Through the SDK
```

Author: I think we should change it to `captureException` to be clear the mechanism we're using.  "SDK ... Through the SDK" is a bit circular.

> When the SDK encounters a "slow click" it will use `sentry.captureException(slowClick)` to create an error issue on the Issues platform.

Reviewer 3: It'll be an envelope with an event in it. We could skip the "Issues Platform" bit?

```suggestion
When the SDK encounters a "slow click" it will use `sentry.captureEvent(slowClick)`.
```

Author: I think we're saying the same things here.  My extra text is confusing.  Happy to remove it.

**Pros:**

- Significantly larger customer reach (i.e. all javascript SDK customers).
- Opportunity to upsell customers on Session Replay from a "slow click" event.
- No code changes required on the Session Replay back end.
- No code changes required by the Issues team.

**Cons:**

- Requires SDK changes.
- Can not change issue sampling parameters without an SDK update.
  - Or can you?
  - What if we sample within the error processing pipeline on the back-end?

>   - If error processing is functionally immutable (i.e. there is no organizational will-power to allow product teams to sample within the error pipeline) then we will not be able to sample on the back-end.

Reviewer 3: ```suggestion
  - If error processing is functionally immutable (i.e. there is no organizational will-power to allow product teams to sample within the error pipeline) then we will not be able to sample on the back-end.
- Consumes from user error quota: If too noisy, might give the impression we're 'wasting' their error quota
```

### Option 2: SDK Creates Issues Through an Issues HTTP Endpoint

> When the SDK encounters a "slow click" it will use `http.post("/issues", data=slowClick)` to create an error issue on the Issues platform.

Reviewer 3: Not sure this is feasible. Given we'd need the whole mechanism of retry/backpressure, serialization, etc. It would need to be a new `envelopeType` and use the same `captureEnvelope` mechanism of the SDK instead.

Author: I'm not sure the exact implementation details.  This was an idea proposed by the Issues team.  A generic /issues endpoint which product teams could make requests to.  My http client psuedo-code shouldn't be interpreted as production code.  I'm only trying to indicate that there's a new /issues HTTP endpoint which is submitted to by the SDK.

Reviewer 3: Would this be rolled out on Relay? How do we handle all the ingestion scaling, PoP's etc? The endpoint could be used internally by the consumer if that helps somehow, but to be called by the clients, I imagine we'll need to stick to `/envelope` and Relay. But happy keep this option here, just raising my concerns.

Author: Not sure of the implementation details.  I would imagine the endpoint would be added to Relay and messages forwarded to the Issues Kafka consumer.

**Pros:**

- Significantly larger customer reach (i.e. all javascript SDK customers).
- Opportunity to upsell customers on Session Replay from a "slow click" event.
- No code changes required on the Session Replay back end.
- We can sample slow click issues on the backend.
  - This would require buy-in and coordination from the Issues team.
  - There is nothing preventing us from looking at an HTTP request and making a go/no-go decision.

> **Cons:**

Reviewer 4: the big problem i see is _noise_. from our tests it seems like dead clicks could drown out error quota quite quickly.

- Requires SDK changes.
- Requies code changes by the Issues team to create a generic interface for creating issues.
  - This would likely disrupt our June 18 deadline.

>   - Unless the Issues team has excess capacity and a willingness to work on it immediately.

Reviewer 4: it also seems like we'd also want to do this in Relay -- we try not to have python endpoints be hit by SDKs. seems like it would be a large amount of work.

### Option 3: Replay SDK Pushes Issues to Session Replay Backend Which Raises an Issue

When the SDK encounters a "slow click" it will append the slow click to the recording events payload. The back-end will search for these events and then publish them to the Issues platform's kafka consumer.

**Pros:**

- We can sample slow click events on the back-end without worrying about input from other teams.
- No code changes required by the Issues team.

> - No SDK changes required as the SDK is already sending slow-click events.

Reviewer 3: We wouldn't need any further changes? I imagine we wouldn't keep it in its current form. For example, not emitting anything for short durations, for example.

If that's true, perhaps the "No SDK change X SDK change" pro/con items could all be removed. Or adjusted with "Complex vs Simple SDK changes" or any other more specific description.

@mydea @billyvg can probably give us the best approach with this

Author: My thinking was "As a result of choosing option X we would or would not need to ask the SDK team to do something".  The SDK team can always make changes to the detection strategy but the outcome of this RFC wouldn't compel changes.

I can remove this text as its a minor point and it seems to cause confusion.

**Cons:**

- Requires the Session Replay SDK.
  - Significantly smaller pool of customers who will see "slow click" issues.

>   - No/Limited opportunity to upsell the Session Replay product.

Reviewer 3: They would already be using Replay, so we could omit this item. one would argue the feature itself is the upsell (Use Replay to get Dead Click Issues):

```suggestion
```

> - Requies code changes by the Session Replay back-end team.

Reviewer 3: ```suggestion
- Requires code changes by the Session Replay back-end team.
```

  - Requires addition of event sampling, issue platform integration, and merging of the replay-event and recording-event payloads.
  - Merging the replay-event and recording-event payloads together is not a trivial change and requires careful deployment.

# Unresolved questions

# Decisions

> We have decided to go with option **3** and couple these new issues to the Session Replay product and use the Session Replay back-end to create issues. We'll very likely need to continue iterating on the detection and be able to iterate on the server gives us a lot more flexibility. The reach will be smaller in terms of total user adoption (require Replay) but the we believe finding the right value to noise ratio is challenging. Having a Replay becomes very important for users to understand what happened and assess impact. 

Reviewer 5: good point, and very true 👍
